File: Grind75/Week1/ValidParentheses.java
class ValidParentheses {
    public boolean isValid(String s) {
        Stack<Character> startBrackets = new Stack<>();
        char[] chars = s.toCharArray();
        for(char c : chars) {

            if(isOpening(c)) {
                startBrackets.push(c);
            }

            else {
                if(startBrackets.isEmpty()) {
                    return false;
                }
                if(!isCorresponding(startBrackets.pop(), c))
                    return false;
            }
        }
        return startBrackets.size() == 0;
    }

    private boolean isCorresponding(char open, char close) {
        if(open == '(') {
            return close == ')';
        } else if(open == '{') {
            return close == '}';
        }
        return close == ']';
    }

    private boolean isOpening(char c) {
        return (c == '(' || c == '{' || c == '[');
    }
}

File: Grind75/Week1/ValidParentheses.py
class ValidParentheses:

    def isValid(self, s: str) -> bool:
        def isOpen(c: str) -> bool:
            return c == '(' or c == '[' or c == '{'

        def corresponding(open: str, close: str) -> bool:
            if open == '(' and  close != ')':
                return False
            if open == '[' and  close != ']':
                return False
            if open == '{' and  close != '}':
                return False
            return True

        stack = []
        for char in s:
            if isOpen(char):
                stack.append(char)
            else:
                if(len(stack) == 0 or not corresponding(stack.pop(), char)):
                    return False
        return len(stack) == 0
